### PR TITLE
Generate dashboard & podmonitor

### PR DESCRIPTION
**Pull request by david-martin**
The changes in this PR are in preparation for deploying monitoring resources (grafana dashboards, pod monitors, alerts etc...) to 2 different enviornments i.e. k8s & openshift.
For more info on why this split is needed, and the new make target, see this doc change https://github.com/Kuadrant/kcp-glbc/compare/main...david-martin:generate-monitoring-resources?expand=1#diff-78502c0b6155bc2963f91d08ca2b0c25ecdf15d7599f7cc4c666686b5dcc3005R1

There should be no obvious difference after this merges.
Installing observability on a local k8s cluster should work as before.
The difference is that the groundwork is in place to allow generating resources for use in openshift from a single source of truth. For example, you will only have to export a dashboard as json and put it in 1 place (config/observability/monitoring_resources/common/dashboard_glbc.json), and the generation step should create the ConfigMap (for a k8s kube-prometheus set up), and a GrafanaDashboard CR (for an openshift w/ observability operator set up) in the right place for either kustomize config.

To verify this change, you can do a `make local-setup`,
and then follow the steps in the local-setup output to install the glbc controller & the observability resources.

### CODE REVIEW

File: config/observability/kubernetes/kustomization.yaml [hunk 1]
```diff
@@ -2,16 +2,7 @@ resources:
   - ../../kube-prometheus
   - ingress-prometheus.yaml
   - ingress-grafana.yaml
-
```
Comment (david-martin) [inline config/observability/kubernetes/kustomization.yaml:5]: the generate target takes care of getting the configmap into the right shape for us now

File: docs/observability/monitoring.adoc [hunk 2]
```diff
@@ -48,50 +22,12 @@ You may also need to specify a ServiceAccount with the `serviceAccountName` fiel
...
-  namespace: openshift-monitoring
-data:
-  config.yaml: |
-    techPreviewUserWorkload:
-      enabled: true
-----
+A suggested approach here is to install the https://github.com/redhat-developer/observability-operator[observability operator], which deploys its own instance of prometheus, prometheus-operator, alertmanager and grafana. This isn't ideal as there is a risk of prometheus-operator CRDs having conflicts with the cluster monitoring stack. However, the prometheus-operator CRDs are relatively stable. It's should be a reasonable stopgap until a better approach is available (such as user defined alerts in Managed Openshift). The observability operator instance of prometheus is smart enough to configure a subset of metrics to federate from the cluster prometheus. This can be useful if you want to use metrics from kube-state or node-exporter.
 
-On OpenShift versions prior to 4.3, or if you do not want to change your cluster monitoring stack configuration, you can refer to the <<Kubernetes>> section in order to deploy a separate Prometheus operator instance.
+There is an example of installing the observability operator included in the Makefile.
```
Comment (david-martin) [inline docs/observability/monitoring.adoc:30]: will be, after this PR lands.

File: Makefile [hunk 3]
```diff
@@ -208,4 +208,72 @@ gen-metrics-docs:
...
 .PHONY: verify-gen-metrics-docs
 verify-gen-metrics-docs: gen-metrics-docs
+	git diff --exit-code
+
+DHALL ?= $(LOCALBIN)/dhall
+DHALL_TO_YAML ?= $(LOCALBIN)/dhall-to-yaml
+DHALL_TO_JSON ?= $(LOCALBIN)/dhall-to-json
+DHALL_TO_YAML_BINARY_LINUX ?= "https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-json-1.7.10-x86_64-linux.tar.bz2"
+DHALL_TO_YAML_BINARY_MACOS ?= "https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-json-1.7.10-x86_64-macos.tar.bz2"
+DHALL_BINARY_LINUX ?= "https://github.com/dhall-lang/dhall-haskell/releases/download/1.41.1/dhall-1.41.1-x86_64-linux.tar.bz2"
```
Comment (maleck13) [inline Makefile:218]: what is the difference between this and the kustomize stuff mike added?

Comment (maleck13) [reply Makefile:218]: I notice we removed some kustomize stuff. Wondering is there overlap between the two tools?

**Review by maleck13 — COMMENTED** [review]
just a couple of comments on use of dhall.

**Comment by david-martin** [comment]
The goal was to allow a single source of truth where you could simply export a dashboard json from grafana, and have that be imported into a k8s cluster for kube-prometheus grafana as a configmap, and also into a GrafanaDashboard CR for grafana-operator in an openshift cluster.
Also, there's a similar goal where prometheus rules (alerts) are defined in a yaml file that you can then unit test with [promtool](https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/). This yaml then needs to be dumped into PrometheusRules CR (under .spec) so the prometheus operator can use it.

kustomize can do the file injection into a configmap with the configmapGenerator.
However, it seems to be limited in allowing you to inject arbitrary files into parts of yaml.
https://github.com/kubernetes-sigs/kustomize/issues/4636
https://github.com/kubernetes-sigs/kustomize/issues/1387

So to solve this we'd need some other tool. 
Dhall can do this, but [yq](https://github.com/mikefarah/yq) could probably do the job fine too.
Here's an example of how a prometheus rules file and a PrometheusRules CR can be generated from dhall
https://github.com/david-martin/dhall-prometheus-operator-alerts/blob/main/alert-rules.dhall
https://github.com/david-martin/dhall-prometheus-operator-alerts/blob/main/alert-rules-cr.dhall
The CR just imports the rules, and both files get generated as yaml https://github.com/david-martin/dhall-prometheus-operator-alerts/tree/main/target

There's a secondary goal here where I'd like to have alert rules in the upstream with sensible defaults.
Then for downstream, we can easily take them and set different thresholds for alerts, or different severity etc..
To do this with kustomize, you'd have to use [patching](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/patches/). Either a yaml file with all the overrides you want,
or a series of json patch statements that set specific fields. e.g. replace the value at this path with this string.
The dhall solution here would probably be using env vars, like a template file, where you want to override something.

So all in all, it can be done with kustomize (and some other file manipulation tool like yq).
The main thing you're missing without dhall is the type checking when its generating the yamls.

**Pull request closed without merging**